src/htcp.cc

```diff
@@ -1346,8 +1398,16 @@ htcpHandleMsg(char *buf, int sz, Ip::Address &from)
     if (!old_squid_format) {
         memcpy(&hdr, hbuf, sizeof(hdr));
     } else {
+        // old Squid format (minor==0) uses a wider struct due to bitfield layout.
+        // Never read more than available; zero-init then copy the safe prefix.
         htcpDataHeaderSquid hdrSquid;
-        memcpy(&hdrSquid, hbuf, sizeof(hdrSquid));
+        memset(&hdrSquid, 0, sizeof(hdrSquid));
+        if ((size_t)hsz >= sizeof(htcpDataHeaderSquid))
+            memcpy(&hdrSquid, hbuf, sizeof(htcpDataHeaderSquid));
+        else
+            // Guaranteed earlier: hsz >= sizeof(htcpDataHeader) (compact prefix).
+            memcpy(&hdrSquid, hbuf, sizeof(htcpDataHeader));
```

Comment:
Please avoid multi-line clauses without braces.

```suggestion
        if ((size_t)hsz >= sizeof(htcpDataHeaderSquid)) {
            memcpy(&hdrSquid, hbuf, sizeof(htcpDataHeaderSquid));
        } else {
            // Guaranteed earlier: hsz >= sizeof(htcpDataHeader) (compact prefix).
            memcpy(&hdrSquid, hbuf, sizeof(htcpDataHeader));
        }
```